astropy/cosmology/funcs/optimize.py

```diff
@@ -1,31 +1,103 @@
 # Licensed under a 3-clause BSD style license - see LICENSE.rst
 """Convenience functions for `astropy.cosmology`."""
 
+from __future__ import annotations
+
 import warnings
+from typing import TYPE_CHECKING, Any, Callable, Protocol, TypedDict, Union
 
 import numpy as np
+import numpy.typing as npt
 
 from astropy.cosmology import units as cu
 from astropy.cosmology.core import CosmologyError
 from astropy.units import Quantity
 from astropy.utils.exceptions import AstropyUserWarning
 
+if TYPE_CHECKING:
+    from scipy.optimize import OptimizeResult
+    from typing_extensions import NotRequired, TypeAlias
```

Comment:
Given that #15686 was merged this should be written as:
```suggestion
if TYPE_CHECKING:
    import sys
    from scipy.optimize import OptimizeResult
    from typing_extensions import TypeAlias
    
    if sys.version_info >= (3, 11):
        from typing import NotRequired
    else:
        from typing_extensions import NotRequired
```

Reply:
Good catch for `sys`.
Adding the Python version check everywhere is going to be so verbose! Let's figure out a better solution...

Reply:
This will be resolved by #15870